crates/lsp/src/lib.rs

```diff
@@ -120,18 +133,184 @@ fn on_initialize(
+    match workspace.into_iter().next() {
+        Some(package) => {
+            let (mut context, crate_id) = prepare_package(package);
+            if check_crate(&mut context, crate_id, false).is_err() {
+                let result = NargoTestRunResult {
+                    id: params.id.clone(),
+                    result: "error".to_string(),
+                    message: Some("The project failed to compile".into()),
+                };
+                return future::ready(Ok(result));
+            };
+
+            let test_funcs = context.get_all_test_functions_in_crate_matching(
+                &crate_id,
+                FunctionNameMatch::Exact(function_name),
+            );
+
+            match test_funcs.into_iter().next() {
+                Some((_, test_function)) => {
+                    #[allow(deprecated)]
+                    let blackbox_solver = acvm::blackbox_solver::BarretenbergSolver::new();
+                    let test_result = run_test(
+                        &blackbox_solver,
+                        &context,
+                        test_function,
+                        false,
+                        &CompileOptions::default(),
+                    );
+                    let result = match test_result {
+                        TestStatus::Pass => NargoTestRunResult {
+                            id: params.id.clone(),
+                            result: "pass".to_string(),
+                            message: None,
+                        },
+                        TestStatus::Fail { message } => NargoTestRunResult {
+                            id: params.id.clone(),
+                            result: "fail".to_string(),
+                            message: Some(message),
+                        },
+                        TestStatus::CompileError(diag) => NargoTestRunResult {
+                            id: params.id.clone(),
+                            result: "error".to_string(),
+                            message: Some(diag.diagnostic.message),
+                        },
+                    };
+                    future::ready(Ok(result))
+                }
+                None => future::ready(Err(ResponseError::new(
+                    ErrorCode::REQUEST_FAILED,
+                    format!("Could not locate test named: {function_name} in {crate_name}"),
+                ))),
+            }
+        }
+        None => future::ready(Err(ResponseError::new(
+            ErrorCode::REQUEST_FAILED,
+            format!("Could not locate package named: {crate_name}"),
+        ))),
+    }
+}
+
+fn on_tests_request(
+    state: &mut LspState,
+    _params: NargoTestsParams,
+) -> impl Future<Output = Result<NargoTestsResult, ResponseError>> {
+    let root_path = match &state.root_path {
+        Some(root) => root,
+        None => {
+            return future::ready(Err(ResponseError::new(
+                ErrorCode::REQUEST_FAILED,
+                "Could not find project root",
```

Comment:
This is duplicated above, would generally recommend grouping these string constants into an enum or something, but not blocking

Reply:
I agree, these need to actually become a custom error type; however, before we can use custom errors, more work needs to happen in async-lsp that neither the maintainer nor I have had time to fix.